WIP: Add draft of CDFW-examples-inspired sampling schemes

paciorek opened the pull request:
This is a draft of how we can give an opinionated set of sampling schemes to `nimbleOccu` users, inspired by what schemes worked well for the CDFW carnivore and bird examples.

I haven't tried running code, and it's not intended to be merged yet.

One main issue is that in the CDFW examples, all latent effects were species-specific random effects. In various block samplers, we blocked such that all coefficients related to a species were sampled together (but in some cases doing this separately for the detection and state parts of the model). With site-specific and obs-specific coefficients, we need to discuss further (probably in a direct conversation). In particular, I need to understand what a site-specific covariate is - I guess this is just a fixed effect that is assumed to be the same across all species? If so it should probably be included in every species-specific block since it would affect the likelihood for each species. 

@kenkellner I think this now sets us up to discuss next steps and possibly for you to start working on next steps. See also some open "TODO" comments in `nimbleOccu.R` where I'd like input from you, in particular on how best to determine particular sets of parameters in the model.

paciorek committed:
Add draft of CDFW-examples-inspired sampling schemes.

paciorek committed:
Make some minor edits to nimbleOccu MCMC config.

kenkellner commented:
@paciorek A couple things as I start to work on this. 

Is it ok for me to directly commit to this branch or do you just want comments?

By "site-specific" I just mean a covariate that has a single value per site e.g. elevation. An obs covariate would be something that varies by both site and repeated sample, e.g. time-of-day or precipitation. For both of these types of covariates, the multispecies model would typically have random slopes by species. So for example `elev` might be a site covariate and `precip` an obs covariate. For the multispecies model the user might give `stateformula = ~elev` and `detformula = ~precip`, but in the process of building the code corresponding random effects by species will be automatically added, so the final occupancy/state formula would be something like `~1 + elev + (1 + elev || speciesID)` and so on.

I think it is pretty unusual for community occupancy models to have some site/obs covariates that are random by species, and others that are not - all coefficients (intercepts and slopes) should all be random by species. So I think at least initially we should assume that is always the case, which simplifies things a bit. This is how unmarked does things and (to my knowledge) also how spOccupancy works. I'd argue the case where the user wants some mixture of random by species and non-random is one where we can assume users write their own code (perhaps using the underlying LINPRED and/or OCCUPANCY macros). Open to other thoughts on this.

The exceptions are species-level covariates, which obviously can't  be random by species, since we only have a single data point per species (e.g., trophic level). So those need to be accounted for for sure. And also perhaps special cases where you have some additional hierarchical structure in the data such as sites being grouped into regions. unmarked and I believe spOccupancy support the latter case to a limited degree. I would suggest we set aside that situation for now and possibly add support for it later.

paciorek commented:
Yes, please feel free to work directly on the branch. If you're able to understand what I was trying to do in terms of identifying variables used in different ways and my indexing of the nodes (one index value per species) in the sampler assignments, it would be great if you just fix things up based on your comments here (which I haven't fully absorbed yet...).

kenkellner committed:
Identify variables for mcmc config (not tested)

kenkellner committed:
configureOccuMCMC generally working now

kenkellner committed:
Fix bug with HMC/slice sampler specification, and put species-specific covariates in PG samplers

kenkellner commented:
@paciorek I think I have this generally working now. I left a couple comments about things I was unsure on. It would be helpful at this point if you were able to quickly look things over and make sure the configurations look as you expect. I've attached a script that quickly test various settings (it's R but github forces me to use txt).

If things look OK to you I will move on to writing some tests.

kenkellner commented:
Here's the correct test file 
[nimbleOccu_tests.txt](https://github.com/user-attachments/files/19674070/nimbleOccu_tests.txt)

paciorek commented:
Great, thanks. It's on my to-do list, but I am on vacation next week and may not get to it until afterwards.

paciorek committed:
Update MCMC configuration for `nimbleOccu`:

 - Add species-level covars to blocks.
 - Fix use of `jointBinary` sampler.
 - Allow `RW_block` for single-species case.
 - Only do slice for sds and noncentered if not HMC.

paciorek commented:
@kenkellner I think I responded to/fixed all the open items, and this is ready for more complete testing (I ran through the basic sanity code in `nimbleOccu_tests.txt` and things looked good. 

In particular, my current thought is that in the various block samplers that we include species-level covariate coeffs in all of the blocks since they are part of the linear predictor for each species.

kenkellner committed:
Update docs

kenkellner committed:
Add tests for single-species model

kenkellner committed:
Tests for multispecies model

kenkellner commented:
@paciorek I think we're good to go here. I'm going to merge this before I move on to other `nimbleOccu` features, ok with you?

paciorek commented:
Yes, glad you proceeded without me. -- Didn't want my previous wording interpreted as sarcasm ;)
